### ISSUE TITLE
Surface Laptop Keyboard Support Status Update

### ISSUE THREAD

**Issue by jasonnewyork** [jakeday/linux-surface#344]
I just installed Ubuntu 18.04 and the keyboard still doesn't work on the Surface Laptop. I found the thread (https://github.com/jakeday/linux-surface/issues/5) but it really didn’t seem like anyone has been consistently successful – and even though it looks like a fix was submitted to the 4.14-rc6 branch it doesn’t look like it made it as I think Ubuntu 18.04 has the 4.15 kernel version. 

What is the path for getting this fix into the release kernel, plus is this thread the right one for installing Jake's patch - looks like some folks had problems with the install.

**Comment by qzed** [comment]
@jasonnewyork 

> I just installed Ubuntu 18.04 and the keyboard still doesn't work on the Surface Laptop.

Yes, support for that is not yet upstream.

> I found the thread (#5) but it really didn’t seem like anyone has been consistently successful

At this point the discussions there are mostly related to custom kernel update/install procedure, as far as I know there are no major issues with keyboard support other than https://github.com/qzed/linux-surfacegen5-acpi/issues/8, but feel free to point out if I'm missing something.

> it looks like a fix was submitted to the 4.14-rc6 branch

I wasn't aware of this, do you have a link?

> What is the path for getting this fix into the release kernel

For the Laptop keyboard support, my current plan is to address the 3 kernel bugs blocking this issue (see https://github.com/qzed/linux-surfacegen5-acpi/tree/master/patches/4.18 patches 1 to 3). I haven't come around yet to discuss those on the mailing list, but I've been in contact with Johan Hovold, one of the serdev-subsystem developers. I've planned to bring up a discussion around serdev related bug (patch 3) on the mailing-list soon, hopefully this weekend. Once this is done I'll need to have a look at the driver dependencies again and then I can prepare the patch for upstream.

> is this thread the right one for installing Jake's patch

If you're still referring to #5, then that's a question I'm asking myself. The problem is I'm not sure where to put these "tech-support" questions. Maybe gitter if anyone would be interested in that?

> looks like some folks had problems with the install

Again, AFAIK most of them were configuration questions for building the kernel from source (which should have been fixed now with a full kernel config available) or other similar things, e.g. now there's still an open bootloader setup issue. They don't really feel like full-blown issues of this repo for me, and I think something more interactive like gitter (or similar) would probably be a better place to handle them. As long as there are people without much Linux experience, those questions will always be there, but we could probably also have one or two looks over the documentation and try to improve things there.

**Comment by jasonnewyork** [comment]
Ok installed your kernel with the patch and I now have the keyboard working with one exception. The patch does not work if you opt to encrypt the filesystem when you install Ubuntu. When this is selected you have to enter a password early in the boot sequence - and at that point where you enter the password the appropriate keyboard driver hasn't been loaded. I worked around this by not enabling encryption (for now), but this will affect others. Do you want me to open a separate issue for this?

**Comment by versipellis** [comment]
I'm not sure if there's a separate issue for it, but I know the keyboard drivers on load are a known issue. Add `hid_multitouch` to `/etc/initramfs-tools/modules` and it should work.

(I came across that trying to do FDE with LUKS when switching from ecryptfs)

**Comment by qzed** [comment]
@jasonnewyork If that doesn't work I'll probably be good to have a new issue for documentation purposes.
